### PR TITLE
add qwen2 classification

### PR DESCRIPTION
**Pull request by ranzhejiang**
- To meet the customer's demands, we need to add the classification function interface of the qwen2 on gaudi
- you can test it with following code
```python
import torch
import habana_frameworks.torch.core as htcore
from optimum.habana.transformers.modeling_utils import adapt_transformers_to_gaudi
from transformers import AutoTokenizer, AutoModel, AutoModelForCausalLM, AutoModelForSequenceClassification, AutoModelForTokenClassification
adapt_transformers_to_gaudi()

tokenizer = AutoTokenizer.from_pretrained("/mnt/HF_models/qwen2-7b", trust_remote_code=True)
model = AutoModelForSequenceClassification.from_pretrained("/mnt/HF_models/qwen2-7b",num_labels=3, trust_remote_code=True).bfloat16().to('hpu')
print(model)

texts = [
"Today’s exercise makes your body feel very relaxed"
]

label_to_id = {
"positive": 0,
"negative": 1,
"neutral ": 2
}

inputs = tokenizer(texts, padding=True, truncation=True, return_tensors="pt")

with torch.no_grad():
outputs = model(**inputs)

logits = outputs.logits
predictions = torch.argmax(logits, dim=-1)

for text, prediction in zip(texts, predictions):
print(f"Text: {text} -> predicted label: {list(label_to_id.keys())[prediction.item()]}")
```

### CODE REVIEW

File: optimum/habana/transformers/models/qwen2/modeling_qwen2.py [hunk 1]
```diff
@@ -1045,3 +1048,221 @@ def apply_customized_rope(q, k, cos, sin, position_ids, training=True):
...
+        return TokenClassifierOutput(
+            loss=loss,
+            logits=logits,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
+
+def fixed_cross_entropy(source, target, num_items_in_batch: int = None, ignore_index: int = -100, **kwargs):
+    reduction = "sum" if num_items_in_batch is not None else "mean"
+    loss = nn.functional.cross_entropy(source, target, ignore_index=ignore_index, reduction=reduction)
```
Comment (Luca-Calabria) [inline optimum/habana/transformers/models/qwen2/modeling_qwen2.py:1233]: why you used "nn.functional.cross_entropy()" rather than "torch.nn.CrossEntropyLoss" from torch 2.6.0?

Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen2/modeling_qwen2.py:1233]: > why you used "nn.functional.cross_entropy()" rather than "torch.nn.CrossEntropyLoss" from torch 2.6.0?

To align with the implement of transformer in https://github.com/huggingface/transformers/blob/94ae1ba5b55e79ba766582de8a199d8ccf24a021/src/transformers/loss/loss_utils.py#L27 . If necessary, we can also replace it with  `torch.nn.CrossEntropyLoss` after testing

**Review by Luca-Calabria — APPROVED** [review]

**Comment by Luca-Calabria** [comment]
LGTM!

**Comment by 12010486** [comment]
Ok, thanks @Luca-Calabria. Waiting for @regisss then

**Comment by ranzhejiang** [comment]
@regisss Would you please review this PR? Thank you.

**Comment by regisss** [comment]
@ranzhejiang There are merge conflicts to solve following #1698.
Can you also make sure your code is up-to-date with Transformers v4.49 please?

Can you also add a test for it? It can be the code snippet you posted aboce :slightly_smiling_face:

**Comment by ranzhejiang** [comment]
Ok, I will solve it later,  and add this demo in example, thanks for review.

**Comment by Luca-Calabria** [comment]
@ranzhejiang @regisss what any update here?

**Comment by regisss** [comment]
My comment above with

> Can you also make sure your code is up-to-date with Transformers v4.49 please?
Can you also add a test for it? It can be the code snippet you posted above 🙂

should be addressed

**Comment by ranzhejiang** [comment]
Sorry for late reply owing to others. For first comment, I have made sure my code is up-to-date with Transformers v4.49, now no confilict, for second comment, I had not find proper tests for this text-classification task example before, so I will add this the code snippet into https://github.com/huggingface/optimum-habana/tree/main/examples/text-classification before Friday, then it will give reference example for our customers.

**Comment by IlyasMoutawwakil** [comment]
hi @ranzhejiang ! hope my question is not self evident, but why do we need to implement those custom task classes ? my understanding is that if they all use the same based model `QwenModel`, which we patch with `QaudiQwenModel`, the original transformers `QwenModelForSequenceClassification` will work just fine, as it's simply a linear layer on top. My question could be interpreted as : what's Gaudi-specific here ?

File: optimum/habana/transformers/models/qwen2/modeling_qwen2.py [hunk 2]
```diff
@@ -1149,3 +1154,221 @@ def apply_customized_rope(q, k, cos, sin, position_ids, training=True):
     else:
         # keep the same implementation as Transformers v4.37.2
         return apply_rotary_pos_emb(q, k, cos[position_ids], sin[position_ids])
+
+class GaudiQwen2ForSequenceClassification(Qwen2ForSequenceClassification):
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        self.model = GaudiQwen2Model(config)
```
Comment (IlyasMoutawwakil) [inline optimum/habana/transformers/models/qwen2/modeling_qwen2.py:1162]: my point is that the line in adapt_transformers_to_gaudi: `transformers.models.qwen2.modeling_qwen2.Qwen2Model = GaudiQwen2Model` already makes these qwen2 models with task heads compatible/optimized for gaudi.

Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen2/modeling_qwen2.py:1162]: What you said make sense, it seems that this plan can not run work well for auto tp when I firstly give this pr, but now it can works well for single card, but I still need confirm what you said in 8-card env.

Comment (ranzhejiang) [reply optimum/habana/transformers/models/qwen2/modeling_qwen2.py:1162]: What you said makes sense, I have changed my code and not  using GaudiQwen2Model for init, and the new PR is in https://github.com/huggingface/optimum-habana/pull/2062
